### PR TITLE
#22: Integration with Kraken API (closes #22)

### LINKED ISSUE #22: Integration with Kraken API
[Project card](https://github.com/buildo/cryptonite/projects/1?fullscreen=true#card-6657146)

## requirements
Get relevant data from the Kraken API and hope that they will be back online

## specs
https://www.kraken.com/help/api

## misc
{optional: other useful info}

### PR DESCRIPTION
**Pull request by MatteoTurri**
Closes #22

Integrates Kraken API

## Test Plan

End to end test

image

> These results can be screenshots, query results, or even just pointers to unit tests that successfully passed.}

> An example of partial test is a field displaying 4 possible values. If 3 values are tested, with screenshots, and 1 is not, then it should be mentioned here.}

### CODE REVIEW

File: api/src/main/scala/cryptonite/kraken/model/KrakenCurrency.scala [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+      case KrakenCurrency.XICN => Some(Currency.ICON)
+      case KrakenCurrency.XLTC => Some(Currency.Litecoin)
+      case KrakenCurrency.XXBT => Some(Currency.Bitcoin)
+      case KrakenCurrency.XXLM => Some(Currency.Stellar)
+      case KrakenCurrency.XXMR => Some(Currency.Monero)
+      case KrakenCurrency.XXRP => Some(Currency.Ripple)
+      case KrakenCurrency.ZEUR => Some(Currency.Euro)
+      case KrakenCurrency.ZGBP => Some(Currency.Pound)
+      case KrakenCurrency.ZUSD => Some(Currency.Dollar)
+      case KrakenCurrency.GNO | KrakenCurrency.KFEE | KrakenCurrency.USDT | KrakenCurrency.XDAO | KrakenCurrency.XMLN |
```
Comment (bytecodeguru) [inline api/src/main/scala/cryptonite/kraken/model/KrakenCurrency.scala]: qui puoi semplicemente scrivere
`case _ => None`
il matching è _match first_, come `switch` in java.

File: api/src/main/scala/cryptonite/kraken/model/KrakenTicker.scala [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package cryptonite.kraken.model
+
+case class KrakenTicker (
+  id: String,
+  a: Float, //"a":["2024.200000","1","1.000"],
+  b: Float //"b":["2019.200000","1","1.000"],
```
Comment (bytecodeguru) [inline api/src/main/scala/cryptonite/kraken/model/KrakenTicker.scala]: questi commenti servono ancora?

File: api/build.sbt [hunk 3]
```diff
@@ -15,6 +15,7 @@ lazy val sttp = "com.softwaremill.sttp" %% "core" % sttpVersion
 lazy val sttpCirce = "com.softwaremill.sttp" %% "circe" % sttpVersion
 lazy val sttpAkka = "com.softwaremill.sttp" %% "akka-http-backend" % sttpVersion
 lazy val catsCore = "org.typelevel" %% "cats-core" % "0.9.0"
+lazy val logback = "ch.qos.logback" % "logback-classic" % "1.2.3"
```
[the remark this answers is not in the document]
Comment (MatteoTurri) [reply api/build.sbt]: L'ha aggiunto Gabro per loggare gli errori di Wiro

File: api/src/main/scala/cryptonite/kraken/KrakenGateway.scala [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+import cats.instances.future._
+import cats.data.EitherT
+
+import cryptonite.model._
+import cryptonite.model.exchange._
+import cryptonite.kraken.model._
+import cryptonite.errors.ApiError
+
+
+class KrakenGateway(implicit ec: ExecutionContext) {
```
[the remark this answers is not in the document]
Comment (MatteoTurri) [reply api/src/main/scala/cryptonite/kraken/KrakenGateway.scala]: Ok 👍

File: api/src/main/scala/cryptonite/kraken/KrakenGateway.scala [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    json.withArray(_.headOption.getOrElse(Json.Null))
+  }
+
+  implicit def krakenResultDecoder[A: Decoder]: Decoder[KrakenResult[A]] =
+    deriveDecoder[KrakenResult[A]]
+      .prepare(_.downField("result").withFocus(objToArray).up)
+
+  implicit val tickerDecoder: Decoder[KrakenTicker] =
+    deriveDecoder[KrakenTicker]
+      .prepare(_.downField("a").withFocus(arrayToFirst).up.downField("b").withFocus(arrayToFirst).up)
```
[the remark this answers is not in the document]
Comment (MatteoTurri) [reply api/src/main/scala/cryptonite/kraken/KrakenGateway.scala]: 😄

**Pull request merged**